src/movies.js

```diff
@@ -1,16 +1,79 @@
-// Iteration 1: All directors? - Get the array of all directors.
-// _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the array of directors. How could you "clean" a bit this array and make it unified (without duplicates)?
+
+
+// Iteration 1: All directors? - Get the moviesay of all directors.
+
+function getAllDirectors(movies){ 
+  const allDirectors = movies.map(function(mov){
+    return mov.director;
+  })
+  return allDirectors;
+}
+
+// _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the moviesay of directors. How could you "clean" a bit this moviesay and make it unified (without duplicates)?
+
+function findUniqueDirectors(movies) {
+  const uniqueDirectors = [...new Set(getAllDirectors(movies))];
+  return uniqueDirectors ;
+}
 
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
 
+function howManyMovies (movies){ 
+  const numberOfDramas = movies.filter(function(movies){
+    return movies.director.includes('Steven Spielberg') && movies.genre.includes('Drama');
+  })
+  return numberOfDramas.length;  
+}
+
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
 
-// Iteration 4: Drama movies - Get the average of Drama Movies
+function ratesAverage(movies){
+   
+  const sumRating = movies.reduce(function(acc, val){
+    if (movies.length === 0) return 0;
+    if (val.rate !== '' && 'rate' in val) { 
+    return acc + val.rate} else {return acc}
+  }, 0);
+  const avg = sumRating / movies.length;
+  return Number(avg.toFixed(2));
+                           
+}
 
+// Iteration 4: Drama movies - Get the average of Drama Movies
+function dramaMoviesRate(movies){
+  const numberOfDramas = movies.filter(function(movies){
+    return movies.genre.includes('Drama');
+  }); 
+  const dramaRates = numberOfDramas.map(function(movies){
+    return movies.rate;
+  }); 
+  const sumOfRates = dramaRates.reduce(function(acc, val){
+  return acc + val;
+  }); 
+  const avg = sumOfRates / dramaRates.length;
+  return Number(avg.toFixed(2)); 
```

Comment:
You could save yourself some code by applying the ratesAverage() here instead or rewriting it's content. If you decide to use this solution, then the other error goes away, because that function already returns 0 when the array is empty. If you decided to keep the code, then you will need to add that additional line.